File: 프로그래머스/코테 연습 모든 문제/가장 먼 노드.java
package com.pcloud.jcodetest;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.*;


@SpringBootTest
class JCodeTestApplicationTests {

    @Test
    public void main() {
        Assertions.assertEquals(3, solution(6, new int[][]{{3,6}, {4,3}, {3,2}, {1,3}, {1,2}, {2,4}, {5,2}}));
    }

    public int solution(int n, int[][] edge) {
        int answer = 0;
        int[] depths = new int[n + 1];
        boolean[][] graph = new boolean[n + 1][n + 1];
        Queue<Integer> nodes = new LinkedList<>();
        nodes.add(1);

        for(int i = 0; i < edge.length; i++) {
            graph[edge[i][0]][edge[i][1]] = true;
            graph[edge[i][1]][edge[i][0]] = true;
        }

        int maxDepth = 0;

        while(nodes.isEmpty() == false) {
            int node = nodes.poll();
            // 1부터 시작이므로 2
            for(int i = 2; i <= n; i++) {
                // depths 계산이 되지 않았고 graph 의 값이 존재할 때.
                if(depths[i] == 0 && graph[node][i]) {
                    // depths[i] = 노드의 깊이 + 1
                    depths[i] = depths[node] + 1;
                    nodes.add(i);
                    maxDepth = Math.max(maxDepth, depths[i]);
                }

            }
        }

        for(int depth : depths) {
            if(depth == maxDepth) {
                answer++;
            }
        }

        return answer;
    }
}

File: 프로그래머스/코테 연습 모든 문제/kotlin/bfs/가장 먼 노드.kt
import java.lang.Math.max
import java.util.*

class Solution {
    fun solution(n: Int, edge: Array<IntArray>): Int {
        var queue: Queue<Int> = LinkedList<Int>()
        var graph = Array(n + 1) { BooleanArray(n + 1 ) { false } }
        var depths = IntArray(n + 1) {0}
        var maxDepth = 0

        for (i in edge.indices - 1) {
            graph[edge[i][0]][edge[i][1]] = true
            graph[edge[i][1]][edge[i][0]] = true
        }

        queue.add(1)

        while(queue.isNotEmpty()) {
            val node = queue.poll()
            for (i in 2..n) {
                if (depths[i] == 0 && graph[node][i]) {
                    depths[i] = depths[node] + 1
                    queue.add(i)
                    maxDepth = max(maxDepth, depths[i])
                }
            }
        }

        return depths.filter { it == maxDepth }.size
    }
}

fun main() {
    Solution().solution(6, arrayOf(intArrayOf(3, 6),
        intArrayOf(4, 3), intArrayOf(3, 2), intArrayOf(1, 3), intArrayOf(1, 2), intArrayOf(2, 4), intArrayOf(5, 2)))
}